feat(perf): continuosly measure on single conn (iperf-style)

mxinden opened the pull request:
Our current throughput tests open a connection, open a stream, up- or download 100MB and close the connection. 100 MB is not enough on the given path (60ms, ~5gbit/s) to exit congestion controller's slow-start. See https://github.com/libp2p/test-plans/issues/261 for details.

Instead of downloading 100MB multiple times, each on a new connection, establish a single connection and continuously measure the throughput for a fixed duration (20s).

Closes #261

marten-seemann reviewed (commented):
Should we do 3 iterations at 20 seconds? Slow start won't take more than 1-2s, so this should give us plenty of time to converge, and it would show situations where the congestion controllers runs into a state that it takes long to recover from (sudden cross-traffic).

marten-seemann commented:
> Preliminary results show fruitful for `https` but not `rust-libp2p/quic`.

Are these results available anywhere?

mxinden commented:
> > Preliminary results show fruitful for `https` but not `rust-libp2p/quic`.
> 
> Are these results available anywhere?

Not yet. Still in a very work-in-progress state.

mxinden commented:
Reaching 4.5 gbit/s with https and >5 gbit/s with rust-libp2p. Still testing. Though this is looking promising.

mxinden commented:
iperf throughput mismatch was due to Nagle's algorithm. Disabled now via `-N`. See previous commit. (Default for the https implementation already https://github.com/golang/go/issues/57530.)

![newplot(10)](https://github.com/libp2p/test-plans/assets/7047859/ee8e81aa-08d8-4c88-9902-82656e555130)

https://observablehq.com/d/682dcea9fe2505c4?branch=perf-exit-slow-start#branch

I still need to investigate more for the other measurements (*-libp2p and quic-go). Will try with a fixed MTU of 1500 next.

//CC @marten-seemann

marten-seemann commented:
> iperf throughput mismatch was due to Nagle's algorithm

Interesting! It's great to see iperf and HTTPS achieving roughly similar results (at least in the limit). This means that our setup is getting more trustworthy!

Looking at the graphs, why are some measurements drawn as boxes and some as points? Why do some have error bars and others don't? The spread seems pretty high, do we need more iterations?

I wouldn't be surprised if quic-go maxed out somewhere around 2 Gbps. At some point, your transfer becomes CPU-limited, depending on the number of kernel offloads that your QUIC stack uses (and that's not the thing we want to benchmark here). That said, I just updated quic-go/perf to quic-go v0.38.1 (https://github.com/quic-go/perf/pull/16, I'll merge the PR once GHA is not broken anymore...), which uses GSO by default. Might be worth rebasing your branch to see if this changes anything.

In go-libp2p, Yamux uses a [16 MB receive window](https://github.com/libp2p/go-libp2p/blob/dce942454b6b85c08fc58867f6e4d5506718e56f/p2p/muxer/yamux/transport.go#L24), which should limit us to roughly 2 Gbps (minus some muxer overhead). It's interesting to see that we're achieving roughly half of that. Could be a coincidence, or point to a bug in our flow control autotuning. I'd be happy to debug this using the current setup (assuming I can still run it manually as I could with the version on master), please let me know.

QUIC uses a [10 MB window](https://github.com/libp2p/go-libp2p/blob/dce942454b6b85c08fc58867f6e4d5506718e56f/p2p/transport/quicreuse/config.go#L13), which limits the bandwidth to 1.25 Gbps. That means we're not quite at the optimum, but pretty close. Would it be helpful for you if we prioritize resolving https://github.com/libp2p/go-libp2p/issues/2290? Alternatively, we could also just have a go-libp2p branch that bumps that value, so we can see if that's actually the root of the problem.

marten-seemann commented:
Here are some interesting result from running the HTTPS test and analyzing the tcpdump. The congestion controller used for this test is Cubic. First interesting result: importing and processing an 8 GB pcap into Wireshark takes a pretty long time O(30min) ;)

Here's the RTT distribution:
<img width="1391" alt="Screenshot 2023-09-07 at 22 59 31" src="https://github.com/libp2p/test-plans/assets/1478487/60006410-8426-4051-8f3d-67955e6f18e4">
There's definitely some queues building up in the network, but it's not too terrible.

Here's the sequence plot (ignore the wrapping of the packet number, obviously), showing the time when packet loss occurred:
<img width="1358" alt="Screenshot 2023-09-07 at 15 51 10" src="https://github.com/libp2p/test-plans/assets/1478487/08b5382b-8fd2-4994-9537-7d0ec9e29adf">

And here's the throughput:
<img width="1391" alt="Screenshot 2023-09-07 at 15 01 25" src="https://github.com/libp2p/test-plans/assets/1478487/b493aea3-1835-445a-b344-215314034058">

Obviously, we're very far from reaching a steady state. 

Here's some  back-of-the-envelope math to calculate the recovery time (i.e. the time it takes to ramp up the congestion window to its original size after a loss), and assuming a BDP (at 5 gbps and 65ms RTT) of roughly 40 MB:
* On Reno, a packet loss halves the cwnd, and every round-trip without a loss event increases it by one MSS. Thus the recovery time is (20 MB / 1400 bytes) ~14000 RTTs, which is 15 minutes (!).
* On Cubic, it's harder to estimate. The [L4S Prague paper](https://www.bobbriscoe.net/projects/latency/tcp-prague-netdev0x13.pdf) claims that at 100 mbps the recovery time is 250 RTTs, and doubles for every 8x increase in bandwidth. That's roughly 1000 RTTs, which is 1 minute.

In the sequence plot above, we see packet loss happening 10x as frequently as this calculation suggests. This might be due to the more shallow buffer, but I don't know precisely how the recovery time scales with the buffer size.

---

What does this mean for our perf setup? At the BDP that we chose for our test, we're running into limitations imposed by the congestion controllers:
* With a recovery time of 15 mintues, Reno has no chance against Cubic whatsoever, yet this is what RFC 9002 recommends for QUIC implementations and what major players have deployed in their QUIC stacks.
* Even Cubic's recovery time is pretty long. At a sampling frequency of 1s, we will pick up the saw-tooth pattern inherent to the congestion controller, and we will inevitably see a wide spread of measurement results.

mxinden commented:
> > Will try with a fixed MTU of 1500 next.
> 
> Does AWS allow larger MTUs on their backbone? That would indeed give TCP an unfair advantage over at least quic-go. Have you verified that using tcpdump / Wireshark?

Turns out, it does not:

```
[ec2-user@ip-]$ ping -M do -s 1472 -c 4 x.x.x.x
PING  () 1472(1500) bytes of data.
1480 bytes from : icmp_seq=1 ttl=109 time=63.4 ms
1480 bytes from : icmp_seq=2 ttl=109 time=63.4 ms
1480 bytes from : icmp_seq=3 ttl=109 time=63.4 ms
1480 bytes from : icmp_seq=4 ttl=109 time=63.4 ms

---  ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3004ms
rtt min/avg/max/mdev = 63.384/63.408/63.441/0.024 ms
```
```
[ec2-user@ip-]$ ping -M do -s 1500 -c 4 x.x.x.x
PING  () 1500(1528) bytes of data.
From  icmp_seq=1 Frag needed and DF set (mtu = 1500)
ping: local error: message too long, mtu=1500
ping: local error: message too long, mtu=1500
ping: local error: message too long, mtu=1500

---  ping statistics ---
4 packets transmitted, 0 received, +4 errors, 100% packet loss, time 3086ms

mxinden commented on perf/README.md:
````diff
@@ -79,11 +79,33 @@ Given you have provisioned your infrastructure, you can now build and run the li
-              ```
-              Note that the measurement includes the time to (1) establish the
-              connection, (2) upload the bytes and (3) download the bytes.
+            - Measurement output is printed to `stdout` as JSON.
+            - The output schema is:
+               ``` typescript
+               interface Data {
+                 type: "intermediary" | "final";
+                 timeSeconds: number;
+                 uploadBytes: number;
+                 downloadBytes: number;
+               }
+               ```
+            - Every second the client must print the current progress to stdout. See example below. Note the `type: "intermediary"`.
+               ``` json
+               {
+                 "type": "intermediary",
+                 "timeSeconds": 1.004957645,
+                 "uploadBytes": 73039872,
+                 "downloadBytes": 0
+               },
+               ```
+            - Before terminating the client must print a final summary. See example below. Note the `type: "final"`. Also note that the measurement includes the time to (1) establish the connection, (2) upload the bytes and (3) download the bytes.
+               ``` json
+               {
+                 "type": "final",
+                 "timeSeconds": 60.127230659,
+                 "uploadBytes": 4382392320,
+                 "downloadBytes": 0
+               }
+               ```
````
This pull request requires the following changes to a perf binaries output format.

This pull request already updates the `go-libp2p` and `https` implementation. https://github.com/quic-go/perf/pull/17 updates the `quic-go` implementation. https://github.com/libp2p/rust-libp2p/pull/4382 updates the rust-libp2p implementation.

The motivation for the larger effort is documented in https://github.com/libp2p/test-plans/issues/261.

@maschad do you have time and interest to do the corresponding changes on the js-libp2p side? If so, as a first step, you could update the existing `stdout` output to the above detailed `type: "final"` output. As a second step the `type: "intermediate"` can follow.

maschad replied:
Thanks for the work on this PR @mxinden and the clear write up. I created an issue https://github.com/libp2p/js-libp2p/issues/2064 to track this and will tackle it once I have completed my current in-flight work. Once that is merged and released I will open a PR to https://github.com/libp2p/test-plans

mxinden commented:
> Looking at the graphs, why are some measurements drawn as boxes and some as points?

The Box visualizes Q1 to Q3 with Q2 (median) denoted with a line within the box. The lines are the whiskers, representing Q0 (minimum) and Q4 (maximum). The dots represent outliers.

https://en.wikipedia.org/wiki/Box_plot has a good explanation for each of these.

> Why do some have error bars and others don't?

I am not sure what you refer to with "error bars" @marten-seemann.

> The spread seems pretty high, do we need more iterations?

I decreased each measurement duration to 20 seconds and increased the iterations per implementation and transport to 10. I will trigger a new CI run soon so that we get more reliable data.

mxinden commented:
> I wouldn't be surprised if quic-go maxed out somewhere around 2 Gbps. At some point, your transfer becomes CPU-limited, depending on the number of kernel offloads that your QUIC stack uses (and that's not the thing we want to benchmark here). That said, I just updated quic-go/perf to quic-go v0.38.1 ([quic-go/perf#16](https://github.com/quic-go/perf/pull/16), I'll merge the PR once GHA is not broken anymore...), which uses GSO by default. Might be worth rebasing your branch to see if this changes anything.

:+1: Note that I merged current `master` into https://github.com/quic-go/perf/pull/17 and updated the reference here. Thus with the next update to `benchmark-results.json` we will see the impact of GSO.

mxinden commented:
> In go-libp2p, Yamux uses a [16 MB receive window](https://github.com/libp2p/go-libp2p/blob/dce942454b6b85c08fc58867f6e4d5506718e56f/p2p/muxer/yamux/transport.go#L24), which should limit us to roughly 2 Gbps (minus some muxer overhead). It's interesting to see that we're achieving roughly half of that. Could be a coincidence, or point to a bug in our flow control autotuning. I'd be happy to debug this using the current setup (assuming I can still run it manually as I could with the version on master), please let me know.

Indeed surprising. You can still run it manually. Please go ahead. Thank you @marten-seemann.

mxinden commented:
I have updated the forked dashboard to the latest data format:

https://observablehq.com/d/682dcea9fe2505c4?branch=27d07a6f47c2bc1a9c9d9a9f6626b536248284f5

mxinden reviewed (commented):
@marten-seemann @sukunrt can either of you give the `perf/impl/https` and `perf/impl/go-libp2p` changes a review?

marten-seemann commented on perf/impl/go-libp2p/v0.29/main.go:
```diff
@@ -99,7 +102,10 @@ func main() {
-	Latency float64 `json:"latency"`
+	Type          string  `json:"type"`
+	TimeSeconds   float64 `json:"timeSeconds"`
+	UploadBytes   uint    `json:"uploadBytes"`
```
This would better be a uint64, otherwise you'll quickly run into limits on 32 bit machines.

marten-seemann commented on perf/impl/go-libp2p/v0.29/perf.go:
```diff
@@ -89,7 +91,26 @@ func sendBytes(s io.Writer, bytesToSend uint64) error {
+
 	for bytesToSend > 0 {
+		now := time.Now()
+		if now.Sub(lastReportTime) > time.Second {
```
```suggestion
		if now.Sub(lastReportTime) >= time.Second {
```

marten-seemann commented on perf/impl/go-libp2p/v0.29/perf.go:
```diff
@@ -100,15 +121,49 @@ func sendBytes(s io.Writer, bytesToSend uint64) error {
 func drainStream(s io.Reader) (uint64, error) {
 	var recvd int64
-	recvd, err := io.Copy(io.Discard, s)
+	recvd, err := io.Copy(io.Discard, & reportingReader { orig: s, LastReportTime: time.Now() })
```
Does it even work with the spaces? Can you run `gofmt ./...` on the code?
```suggestion
	recvd, err := io.Copy(io.Discard, &reportingReader{ orig: s, LastReportTime: time.Now() })
```

marten-seemann commented on perf/impl/go-libp2p/v0.29/perf.go:
```diff
@@ -100,15 +121,49 @@ func sendBytes(s io.Writer, bytesToSend uint64) error {
+	r.lastReportRead += uint64(n)
+
+	now := time.Now()
+	if now.Sub(r.LastReportTime) > time.Second {
```
```suggestion
	if now.Sub(r.LastReportTime) >= time.Second {
```

marten-seemann reviewed (commented):
I reviewed the go-libp2p implementation.

mxinden commented:
The last commit removes js-libp2p. Once https://github.com/libp2p/js-libp2p/pull/2067 is merged, we can re-introduce it here.
